digdag-standards/src/main/java/io/digdag/standards/operator/td/TdOperatorFactory.java

```diff
@@ -78,46 +85,126 @@ public Operator newTaskExecutor(Path workspacePath, TaskRequest request)
-        }
 
-        @Override
-        public TaskResult runTask()
-        {
-            Config params = request.getConfig().mergeDefault(
+            this.params = request.getConfig().mergeDefault(
                     request.getConfig().getNestedOrGetEmpty("td"));
 
-            String query = templateEngine.templateCommand(workspacePath, params, "query", UTF_8);
+            this.query = templateEngine.templateCommand(workspacePath, params, "query", UTF_8);
 
-            Optional<TableParam> insertInto = params.getOptional("insert_into", TableParam.class);
-            Optional<TableParam> createTable = params.getOptional("create_table", TableParam.class);
+            this.insertInto = params.getOptional("insert_into", TableParam.class);
+            this.createTable = params.getOptional("create_table", TableParam.class);
             if (insertInto.isPresent() && createTable.isPresent()) {
                 throw new ConfigException("Setting both insert_into and create_table is invalid");
             }
 
-            int priority = params.get("priority", int.class, 0);  // TODO this should accept string (VERY_LOW, LOW, NORMAL, HIGH VERY_HIGH)
-            Optional<String> resultUrl = params.getOptional("result_url", String.class);
+            this.priority = params.get("priority", int.class, 0);  // TODO this should accept string (VERY_LOW, LOW, NORMAL, HIGH VERY_HIGH)
+            this.resultUrl = params.getOptional("result_url", String.class);
 
-            int jobRetry = params.get("job_retry", int.class, 0);
+            this.jobRetry = params.get("job_retry", int.class, 0);
 
-            String engine = params.get("engine", String.class, "presto");
+            this.engine = params.get("engine", String.class, "presto");
 
-            Optional<String> downloadFile = params.getOptional("download_file", String.class);
+            this.downloadFile = params.getOptional("download_file", String.class);
             if (downloadFile.isPresent() && (insertInto.isPresent() || createTable.isPresent())) {
                 // query results become empty if INSERT INTO or CREATE TABLE query runs
                 throw new ConfigException("download_file is invalid if insert_into or create_table is set");
             }
 
-            boolean storeLastResults = params.get("store_last_results", boolean.class, false);
+            this.storeLastResults = params.get("store_last_results", boolean.class, false);
+
+            this.preview = params.get("preview", boolean.class, false);
+
+            this.state = request.getLastStateParams().deepCopy();
+
+            this.existingJobId = state.getOptional(JOB_ID, String.class);
+            this.existingDomainKey = state.getOptional(DOMAIN_KEY, String.class);
+        }
+
+        @Override
+        public TaskResult runTask()
+        {
+            try (TDOperator op = TDOperator.fromConfigWithRetries(params)) {
+
+                // Generate and store domain key before starting the job
```

Comment:
Maybe there should be a way to persist task state without having to exit / throw `TaskExecutionException`. 

Maybe we should inject an interface that the operator can use to store state directly. This would avoid the extra `TaskExecutionException` and the associated overhead of querying the DB to update the task state etc.